modyn/supervisor/internal/triggers/model_downloader/model_downloader.py

```diff
@@ -0,0 +1,106 @@
+
+logger = logging.getLogger(__name__)
+
+
+# TODO(366) Unify similar code in trainer_server and evaluator. Create common.utils.ModelDownloader
+class ModelDownloader:
+    def __init__(
+        self,
+        modyn_config: dict,
+        pipeline_id: int,
+        device: str,
+        base_dir: pathlib.Path,
+        model_storage_address: str,
+    ):
+        self.modyn_config = modyn_config
+        self.pipeline_id = pipeline_id
+        self._device = device
+        self._device_type = "cuda" if "cuda" in self._device else "cpu"
+        self.base_dir = base_dir
+        assert self.base_dir.exists(), f"Temporary Directory {self.base_dir} should have been created."
+
+        self._model_storage_stub = self.connect_to_model_storage(model_storage_address)
+
+        self.model_configuration_dict: Optional[dict] = None
+        self.model_handler = None
+        self._amp: Optional[bool] = None
+        self._model = None
+
+    @staticmethod
+    def connect_to_model_storage(model_storage_address: str) -> ModelStorageStub:
+        model_storage_channel = grpc.insecure_channel(model_storage_address)
+        assert model_storage_channel is not None
+        if not grpc_connection_established(model_storage_channel):
+            raise ConnectionError(
+                f"Could not establish gRPC connection to model storage at address {model_storage_address}."
+            )
+        return ModelStorageStub(model_storage_channel)
+
+    def _load_state(self, path: pathlib.Path) -> None:
+        assert path.exists(), "Cannot load state from non-existing file"
+        assert self._model is not None
+
+        logger.info(f"Loading model state from {path}")
+        with open(path, "rb") as state_file:
+            checkpoint = torch.load(io.BytesIO(state_file.read()), map_location=torch.device("cpu"))
+
+        assert "model" in checkpoint
+        self._model.model.load_state_dict(checkpoint["model"])
+
+        # delete trained model from disk
+        path.unlink()
+
+    def configure(self, model_id: int) -> None:
+        with MetadataDatabaseConnection(self.modyn_config) as database:
+            trained_model: Optional[TrainedModel] = database.session.get(TrainedModel, model_id)
+            if not trained_model:
+                logger.error(f"Trained model {model_id} does not exist!")
+                return
+            model_class_name, model_config, amp = database.get_model_configuration(trained_model.pipeline_id)
+        self._amp = amp
+        model_module = dynamic_module_import("modyn.models")
+        self.model_handler = getattr(model_module, model_class_name)
+        assert self.model_handler is not None
+
+        self.model_configuration_dict = json.loads(model_config)
+        self._model = self.model_handler(self.model_configuration_dict, self._device, amp)
+
+        assert self._model is not None
+        assert isinstance(self._model.model, CoresetSupportingModule)
```

Comment:
Why this assert? I think not every model has to be a coreset supporting moudle, or?

Reply:
Now, only the data drift trigger uses the model downloader, so the models should be able to record embeddings. Will deal with it in the general model downloader PR.